Add LLaVa 1.6

NielsRogge opened the pull request:
# What does this PR do?

This PR adds the new LLaVa 1.6 model.

To do:

- [x] not sure how batched generation works => is supported in case `image_sizes` are the same for all images
- [x] make `image_sizes` a tensor instead of a list
- [x] make sure llava 1.5 still works

NielsRogge committed:
First draft

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Improve conversion script

NielsRogge committed:
Improve conversion script

NielsRogge committed:
Improve script

NielsRogge committed:
Update script

NielsRogge committed:
More improvements

NielsRogge committed:
Convert logits

NielsRogge committed:
Add generation

NielsRogge committed:
More improvements

NielsRogge committed:
More improvements

NielsRogge committed:
Make image_sizes a tensor

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29012). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

NielsRogge committed:
Add support for batched generation

NielsRogge committed:
Use appropriate prompt

NielsRogge committed:
More improvements

NielsRogge committed:
Make fixup

NielsRogge committed:
Fix docstrings

NielsRogge committed:
Improve conversion script

NielsRogge committed:
More improvements

NielsRogge committed:
Make fixup

NielsRogge committed:
Debug

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_llava_1_6

NielsRogge committed:
More improvements

NielsRogge committed:
Support padding of image

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_llava_1_6

NielsRogge committed:
Remove unused image_aspect_ratio

NielsRogge committed:
Add use_image_newline_parameter

NielsRogge committed:
More improvements

NielsRogge committed:
Remove script

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_llava_1_6

NielsRogge committed:
Add integration test

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_llava_1_6

amyeroberts commented on src/transformers/models/llava/convert_llava_1_6_to_hf.py:
```diff
@@ -0,0 +1,266 @@
+    # Pad to 64 for performance reasons
+    pad_shape = 64
+    model.resize_token_embeddings(config.text_config.vocab_size + 2, pad_shape)
+    model.language_model.model.embed_tokens.weight.data[32000:] = torch.stack(
+        tuple((dist.sample() for _ in range(model.language_model.model.embed_tokens.weight.data[32000:].shape[0]))),
+        dim=0,
+    )
+    model.language_model.lm_head.weight.data[32000:] = torch.stack(
+        tuple((dist.sample() for _ in range(model.language_model.lm_head.weight.data[32000:].shape[0]))),
+        dim=0,
+    )
```
Is we're sampling from `dist` here then we should set a seed

NielsRogge replied:
cc @ArthurZucker I took this from the original llava conversion script

amyeroberts commented on src/transformers/models/llava/convert_llava_1_6_to_hf.py:
```diff
@@ -0,0 +1,266 @@
+        dim=0,
+    )
+
+    device = "cuda:3"
```
Let's use a different device selection - not everyone will have 3 GPUs

amyeroberts commented on src/transformers/models/llava/image_processing_llava.py:
```diff
@@ -0,0 +1,616 @@
+
+def select_best_resolution(original_size: tuple, possible_resolutions: list) -> tuple:
+    """
+    Selects the best resolution from a list of possible resolutions based on the original size.
```
This should explain what it means for something to be the "best" resolution

NielsRogge replied:
I've expanded the docstring

amyeroberts commented on src/transformers/models/llava/image_processing_llava.py:
```diff
@@ -0,0 +1,616 @@
+        ):
+            max_effective_resolution = effective_resolution
+            min_wasted_resolution = wasted_resolution
+            best_fit = (width, height)
```
We should be consistently working with (h, w) format

amyeroberts commented on src/transformers/models/llava/image_processing_llava.py:
Tests should be added for the image processor - especially for logic which is used in the model's forward pass like `select_best_resolution`

NielsRogge replied:
I've added tests for this

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -409,25 +477,79 @@ def forward(
-                    selected_image_feature = selected_image_feature[:, 1:]
-                elif vision_feature_select_strategy == "full":
-                    selected_image_feature = selected_image_feature
+                if pixel_values.ndim == 5:
```
There should be a comment here explaining what `ndim == 5` represents

NielsRogge replied:
I've added the shapes in the line below

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -409,25 +477,79 @@ def forward(
-                elif vision_feature_select_strategy == "full":
-                    selected_image_feature = selected_image_feature
+                if pixel_values.ndim == 5:
+                    concat_images = torch.cat(list(pixel_values), dim=0)
```
Why are we converting the pixel values to a list here?

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -409,25 +477,79 @@ def forward(
+                        selected_image_feature = selected_image_feature[:, 1:]
+                    elif vision_feature_select_strategy == "full":
+                        selected_image_feature = selected_image_feature
+                    else:
+                        raise ValueError(
+                            f"Unexpected select feature strategy: {self.config.vision_feature_select_strategy}"
+                        )
```
Rather than have this check in two different places - it would be better to validate if `vision_feature_select_strategy` is valid at the start of the method: fail early

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -409,25 +477,79 @@ def forward(
+                    split_sizes = [image.shape[0] for image in pixel_values]
+                    image_features = torch.split(image_features, split_sizes, dim=0)
+
+                    # NOTE we only support multimodal_patch_merge_type == "spatial_unpad"
```
This should be asserted somewhere and added to a docstring - most users won't read the code here

NielsRogge replied:
This doesn't need to be asserted since `multimodal_patch_merge_type` isn't a configurable parameter, I've removed it since this is the only patch merging type that the authors eventually used. Hence the comment which I added

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -46,6 +49,62 @@
+    Returns:
+        tuple: The shape of the image patch grid in the format (width, height).
+    """
+    if isinstance(grid_pinpoints, list):
+        possible_resolutions = grid_pinpoints
+    else:
+        possible_resolutions = literal_eval(grid_pinpoints)
```
This is going to be called repeatedly in the forward pass. It's better to properly parse the grid points once in the model's init and then pass that to the method

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -240,6 +299,10 @@ def __init__(self, config: LlavaConfig):
         self.multi_modal_projector = LlavaMultiModalProjector(config)
+
+        if config.use_image_newline_parameter:
+            self.image_newline = nn.Parameter(torch.empty(config.text_config.hidden_size, dtype=self.dtype))
```
What was the conclusion about adding parameters in the other model PR discussion?

NielsRogge replied:
This PR proposes to add support for llava 1.6 in modeling_llava.py. The only new trainable parameter is this one

amyeroberts commented on src/transformers/models/llava/modeling_llava.py:
```diff
@@ -409,25 +477,79 @@ def forward(
+
+                    # NOTE we only support multimodal_patch_merge_type == "spatial_unpad"
+                    new_image_features = []
+                    for image_idx, image_feature in enumerate(image_features):
```
This is going to be super slow iterating over the image like this. Looking at the code, it looks like this can all be vectorized

NielsRogge replied:
I'm afraid this can't be vectorized, since you need to go over each of the individual image features since the image sizes differ per image in the batch. This was taken from the original implementation here: https://github.com/haotian-liu/LLaVA/blob/5d8f1760c08b7dfba3ae97b71cbd4c6f17d12dbd/llava/model/llava_arch.py#L166

ArthurZucker replied:
the goal is to not follow the original implementation. Which means more work for use of course but we did not follow the original implementation either and successfully vectorized the forward pass 😉

NielsRogge replied:
Could both of you elaborate a bit on how this code can be vectorized? Because the `images_sizes` can be different for each image in the batch, so I'm not seeing how this can be easily vectorized

NielsRogge replied:
Basically this line differs per image in the batch, preventing batching: https://github.com/NielsRogge/transformers/blob/add_llava_1_6/src/transformers/models/llava/modeling_llava.py#L508

ArthurZucker replied:
I have no idea what this is doing so I don't know, but maybe we can change the paradigm in a way that prevents us from having to do this. 
Think it that way: if the processor has the number of images, the size of the images and the raw text, it can act on a text level, which is super super fast even if you do it sequentially. 
If the pre-processing is just giving more embedding tokens to images (so a bigger image has 512 + more tokens, while a smaller image has 512 - less tokens) then we can just do that in the processor. Replace the `<image>` token with `num_image_tokens[i] * "<fake_image_token"` which should be alot lot simpler

nxphi47 replied:
@NielsRogge @ArthurZucker Hey, nice works here. I'm re-implementing your model code so that that this unpadding steps will be more efficient.
It's true that you can't vectorize the images completely, but you can do so by pre-calculating the indices and the final size of the full_image_features (for all images). Here is the steps:

- From the image_sizes, use your for loop to calculate how many tokens each image has, how many patches (it's not always 5) each image has, and the slice indices of each patche (incase of unpadding). This can be done with cpu because you don't deal with tensor here.
- Create a single `full_image_embeds` of size (all_img_tokens, embed_dim).
- for loop through image features and copy the indexed img features to the correct position of `full_image_embeds`. (*This will be faster than your implementation with `torch.cat` as it will create new memory each time*, and this copying can be done asynchronously by GPU)
- finally, with the indices, copy `full_image_embeds` to `final_embeddings` where <image> is.

amyeroberts replied:
Awesome - thanks for sharing @nxphi47! Do you have any numbers on how much this speeds things up?

NielsRogge replied:
Hi @nxphi47 if you're willing to open a PR regarding this - that'd be great! I think this could also solve batched generation which is requested at #29832

nxphi47 replied:
@NielsRogge Working on it. My implementation is based on your previously cancelled PR, so I'm adapting it to the new llava-next. 

As for #29832 , the idea is concatenating pixel_values along patch dim (e.g `(8, 3, 336, 336)` for 2 images with patch length 5 and 3), and not stacking into a batch dimension.

NielsRogge replied:
Ok so based on offline discussion with @ArthurZucker, the idea is that this could already be handled in the tokenizer (`LlavaNextProcessor`) without the model requiring the `image_sizes` argument. Basically there we can already calculate the amount of image tokens per image in a batch, and add them to the `input_ids`. 

The model will then replace the special image tokens in the `input_ids` by the image features.

Is this similar to what you were thinking of?

nxphi47 replied:
@NielsRogge cannot do that without image_sizes, because you need to inject the newline_embedding between patches and unpad the patch vision feature after vision_tower.

ArthurZucker replied:
What if you: 
1. Process the images and get the `image_sizes`
2. Process the strings, replacing fake tokens with `image_sizes[batch][image]` image tokens + `new_line`
then the new_line embedding will be correctly placed no?

NielsRogge replied:
Yeah I was also think that if you add the `new_line` token to the `input_ids`, then in the model this could be replaced by the `image_newline` trainable parameter

nxphi47 replied:
@NielsRogge @ArthurZucker  You still have to handle unpadding of image features.

Anyway, I port my implementation here: https://github.com/huggingface/transformers/pull/29850

I have not done testing and bugs are expected, but this should essentially tell you how I'm doing it.

It supports:
- Batched forward with multiple image of different sizes (number of patches).
- Training, for cases without any image.
- multi-image in same sequence. e.g: `[" <image> <image> the first image is a dog while the second is a cat", "    <image> <image> <image> <image these 4 image are..."]`

amyeroberts reviewed (commented):
Thanks for all the work adding this! 

General comments:
* Image processor needs the most work. It breaks a lot of conventions - in particular using (w, h) format and PIL images - which need to be resolved before merging
* Tests should be added both for the new model addition and for the image processor

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge commented on src/transformers/models/llava/convert_llava_1_6_to_hf.py:
```diff
@@ -0,0 +1,266 @@
+    if model_id == "liuhaotian/llava-v1.6-mistral-7b":
+        expected_text = '[INST]  \nWhat is shown in this image? [/INST] The image appears to be a radar chart, which is a type of multi-dimensional plot that displays data in the form of a two-dimensional chart of three or more quantitative variables represented on axes starting from the same point.\n\nIn this particular radar chart, there are several axes labeled with different metrics or benchm ...
+    elif model_id == "liuhaotian/llava-v1.6-vicuna-7b":
+        expected_text = """A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human\'s questions. USER:  \nWhat is shown in this image? ASSISTANT: The image appears to be a graphical representation of a benchmarking study comparing the performance of various models or systems. It\'s a scatter plot with a circul ...
```
Yes, they require a specific chat template.

NielsRogge committed:
Address comments

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_llava_1_6

NielsRogge commented on src/transformers/models/llava/convert_llava_1_6_to_hf.py:
```diff
@@ -0,0 +1,266 @@
+
+    # verify inputs
+    if model_id == "liuhaotian/llava-v1.6-mistral-7b":
+        # replace -200 by 32000 (ask Arthur)
```
Because @ArthurZucker decided to use a positive integer rather than -200

NielsRogge committed:
Address comment

NielsRogge committed:
Make image_sizes height width

NielsRogge committed:
Improve _preprocess

NielsRogge committed:
Address comments

NielsRogge committed:
Address comment

NielsRogge committed:
Rename attribute

NielsRogge committed:
Address comment

NielsRogge committed:
Use height width everywhere

NielsRogge commented on src/transformers/models/llava/image_processing_llava.py:
```diff
@@ -0,0 +1,616 @@
+        image_mean = image_mean if image_mean is not None else self.image_mean
+        image_std = image_std if image_std is not None else self.image_std
+        do_convert_rgb = do_convert_rgb if do_convert_rgb is not None else self.do_convert_rgb
+
```
Added

NielsRogge committed:
Use pad

NielsRogge committed:
Rename variables

NielsRogge committed:
Add image processor tests

NielsRogge committed:
Improve tests

NielsRogge committed:
Improve image processor

NielsRogge committed:
Update modeling

NielsRogge committed:
Address comment

NielsRogge committed:
Address comment

NielsRogge commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -28,20 +28,20 @@
 
     attributes = ["image_processor", "tokenizer"]
-    image_processor_class = "CLIPImageProcessor"
+    image_processor_class = "AutoImageProcessor"
```
Ok, added a `ValueError` similar to the PR above

NielsRogge committed:
Make fixup

amyeroberts commented on src/transformers/models/llava/image_processing_llava.py:
```diff
@@ -0,0 +1,672 @@
+            raise ValueError(f"Invalid aspect ratio setting: {aspect_ratio_setting}")
+
+        new_images = []
+        image_sizes = [get_image_size(image) for image in images]
```
Should use input_data_format here

amyeroberts commented on src/transformers/models/llava/image_processing_llava.py:
```diff
@@ -0,0 +1,672 @@
+
+        return padded_image
+
+    def get_image_patches(self, image: np.array, grid_pinpoints, size: tuple, input_data_format) -> List[np.array]:
```
Should also accept `data_format` as an argument

amyeroberts reviewed (commented):
Thanks for iterating on adding this model! 

The image processor still needs a bit of work. Namely, numpy equivalents of PIL methods should be written instead of converting back and forth between numpy and PIL which is very inefficient and also causes breaks in assumptions about the data format of the images as they go through the processor. 

Please make sure the look over the PR before asking for review - there's a file which should be included in the PR. 

Modified forward pass looks OK to me in terms of logic but does make more complex than before. Would be good to have @ArthurZucker's input there if it's OK.

NielsRogge committed:
Address comments

NielsRogge committed:
Add resample and input_data_format

NielsRogge committed:
Address comments

NielsRogge committed:
Fix image processor tests

NielsRogge committed:
Add data_format

NielsRogge committed:
Add type hints

NielsRogge committed:
Remove script

NielsRogge commented:
Thanks, I've addressed all comments!

Will update the conversion script to use cuda:0 when all checkpoints are uploaded (34b models still need to be uploaded)

ArthurZucker reviewed (commented):
- 🔴 changes to the `LlavaForCausalLM` are just to big to say it's still Llava. We can either have a new forward, or we do what we always do, add a new "Architecture" to properly seperate both. I don't mind having it in the same file, but the expected input is different from normal Llava. 
- 🔴 changes are not explained in the slightest, so coming from the llava code I wrote I have no idea what is happening, why we need to iterate of the images etc. This needs to be explained and needs a separate code-path
- 🟢 on pretty much the other changes (processor tests etc) and leave it to @amyeroberts for them!

NielsRogge committed:
Test batched generation

NielsRogge commented:
Ok thanks for your review.

> 🔴 changes to the LlavaForCausalLM are just to big to say it's still Llava. We can either have a new forward, or we do what we always do, add a new "Architecture" to properly seperate both. I don't mind having it in the same file, but the expected input is different from normal Llava.

I feel like adding a whole new architecture is a bit of an overkill, as the only change of llava 1.6 is that pixel values are now of shape 5 instead of 4 (batch_size, num_patches, num_channels, height, width). Llava 1.6 is also still part of the same paper, they just extend it to work on arbitrary image resolutions.

Could you clarify what you mean by a new forward, since the model LlavaForConditionalGeneration can only have a single forward method defined, I assume. Do you mean defining a whole new class as well?

> 🔴 changes are not explained in the slightest, so coming from the llava code I wrote I have no idea what is happening, why we need to iterate of the images etc. This needs to be explained and needs a separate code-path

I'll add some more comments to explain what's happening.

NielsRogge commented:
Closing this PR in favor of #29586
